lib/child_process.js

```diff
@@ -176,17 +174,26 @@ exports.execFile = function(file /*, args, options, callback*/) {
 
     if (!callback) return;
 
-    // merge chunks
-    var stdout;
-    var stderr;
-    if (!encoding) {
-      stdout = Buffer.concat(_stdout);
-      stderr = Buffer.concat(_stderr);
-    } else {
-      stdout = _stdout;
-      stderr = _stderr;
+    var stdout = Buffer.concat(_stdout, stdoutLen);
+    var stderr = Buffer.concat(_stderr, stderrLen);
+
+    var stdoutEncoding = encoding;
+    var stderrEncoding = encoding;
+
+    if (stdoutState && stdoutState.decoder && stdoutState.decoder.encoding) {
```

Comment:
The `std*State` checks can be removed, they will always evaluate to "true." Same with `decoder.encoding`, if a decoder instance is set, it will always have an `.encoding`.